File: 2018/nim/day1/day1.nim
import math
import sequtils
import sets
import strutils

proc parse_freq_change(s: string): int =
  let
    sign = if s[0] == '+': 1
           else: -1
    number = parseInt(s[1 .. s.high])
    
  sign * number

proc qb(freqs: seq[int]): int =
  var
    fchange: int
    freq = 0
    index = 0
    length = freqs.high + 1
    found = false
    known_freqs = toSet([freq])

  while not found:
    fchange = freqs[index]
    freq += fchange
    if known_freqs.contains(freq):
      found = true
    else:
      index = (index + 1).mod(length)
      known_freqs.incl(freq)

  freq

var freqs = readAll(stdin).strip().splitLines().map(parse_freq_change)
echo("qa: ", freqs.sum())
echo("qb: ", qb(freqs))

File: 2018/python3/day1/day1.py
from itertools import cycle
from sys import stdin

def qb(freqs):
    freq = 0
    known_freqs = set([freq])
    for fchange in cycle(freqs):
        freq += fchange
        if freq in known_freqs:
            return freq
        else:
            known_freqs.add(freq)

if __name__ == '__main__':
    freqs = list(map(lambda line: int(line.strip()), stdin.readlines()))
    print(f"qa: {sum(freqs)}")

    print("qb: {}".format(qb(freqs)))
